add uc_hook_set_user_data

PhilippTakacs opened the pull request:
This allows to change the user_data for hooks. This way it's simpler to adopt the hooks for current needs. I.e. change the page table entries for the tlb hook.

PhilippTakacs committed:
add uc_hook_set_user_data

This allows to change the user_data for hooks. This way it's simpler to
adopt the hooks for current needs. I.e. change the page table entries
for the tlb hook.

wtdcode commented:
LGTM.

PhilippTakacs commented:
> LGTM.

Shall I rebase this to current dev to rerun the CI or is there anything else blocking the merge?

wtdcode commented on uc.c:
```diff
@@ -2060,6 +2060,17 @@ uc_err uc_hook_del(uc_engine *uc, uc_hook hh)
+uc_err uc_hook_set_user_data(uc_hook hh, void *user_data)
+{
+    struct hook *hook = (struct hook *)hh;
+    if (hook->type == UC_HOOK_BLOCK || hook->type == UC_HOOK_CODE) {
```
Wait why do we exclude these two types?

PhilippTakacs replied:
Because the callbacks are directly called by the tb with the user_data pointer written to the tb. It would require to rebuild the tb to allow change the user_data.

wtdcode replied:
Oh that's correct. Can we inline the pointer to the hook struct in the tb instead? In that case, the modification to `uc_hook` would take effect as well.

PhilippTakacs replied:
No because there is no wrapper function in unicorn for these hooks. The cb function is directly called. I could write a wrapper function which just use the hook_struct. This might affect performance which I would like to avoid in the block case.

wtdcode replied:
Oh okay, I see. I forgot the inline optimization.

A wrapper would hurt performance quite a lot. Spare me some time to think about if we have better ways.

PhilippTakacs replied:
What could be done is only allow to change the user_data of code/block hooks while unicorn isn't running and then clear the affected tb cache. This would only affect performance when you change the user data.

PhilippTakacs replied:
A I missed that I have designed the API without an uc_engine, so I can't test if the emulation is running. So I would say it's ok to just disallow changing user data for these hooks. What do you think?

wtdcode replied:
What do you think of registering the data pointers as TCG registers? Like how we sync `env->pc`, we could keep the data pointer updated every time before calling the inline callbacks. This imposes the minimum overhead (a pointer read).

Not a compulsory change, and we can keep the current behavior for simplicity.

wtdcode replied:
Another point: according to https://github.com/unicorn-engine/unicorn/blob/c24c9ebe773ce6fbecb0e39f68ffb23b7326b17f/qemu/include/tcg/tcg-op.h#L50

It is still possible to change the data of the code/block hooks if there are more than 1 hook. Maybe it still makes sense to have the API with `uc_engine` and clear tb cache instead?

PhilippTakacs replied:
Ok let's split this up:

1. changing the hook data while emulation is not running:

This is quite simple just add a uc_engine and clear the affected tb cache. I'll add this.

2. changing the hook data while emulation is running

This is quite tricky.

The code you linked is for tb generation, so it is not usable during emulation. Btw. I think the `==` is wrong and should be a `>=`.

Syncing the userdata like `env->pc` just adds a register read. The problem I see is that the user_data doesn't change this often and changing the user_data of code/block hooks during emulation is not the feature everyone needs. Also this adds complex code. So because I believe this is rarely used I would try to avoid changing the tb generation code.

The third option is to handle this is handle it like self modifying code (see #2271 and #2279), so clear the tb cache and break the translation loop. This is also quite complex and has some quirks on using it (mostly block rerunning).

So all in all I would say the best way to handle this is not to allow changing the user data of code/block hooks while emulation is running.

wtdcode replied:
The logic is correct by using == 1, i.e., if there is exactly one code hook we do inline optimisations where changing data is not possible. When we have more hooks, the code here generates a uc_tracecode helper that iterates all hooks so that changing data is rather simple.

If uc is passed with hooks, as far as I can tell, the only invalid case is only having 1 code hook due to inline optimisation . Or would you like to enable inlining for all cases?

wtdcode replied:
For a bit more context, not enabling inlining for more than 1 hooks was due to a bad hack in tcg_optimize leading to segfault at that time. Since I removed the hack recently, we could extend inlining to any number of hooks.

PhilippTakacs replied:
I somehow misinterpreted what this code is doing, thanks for explaining.

wtdcode replied:
Also I would also recommend not allowing changing data during emulation.

PhilippTakacs replied:
So I guess this can be merged now?

wtdcode replied:
yeah, look good now except the small typo and documents above.

PhilippTakacs commented:
@wtdcode can you merge this PR or is there something missing/blocking?

PhilippTakacs committed:
hooks allow to change the user data of block and code hooks

Now it is possible to change the user data of block and code hooks. The
corresponding translation blocks are cleared. It is only allowed to be
done while the emulation is stopped.

PhilippTakacs committed:
fixup! hooks allow to change the user data of block and code hooks

wtdcode commented on include/unicorn/unicorn.h:
```diff
@@ -1149,6 +1149,22 @@ uc_err uc_hook_add(uc_engine *uc, uc_hook *hh, int type, void *callback,
+ @user_data: user-defined data. This will be passed to callback function in its
+      last argument @user_data
+
+ @return UC_ERR_OK on success, or UC_ERR_ARGS when the hook was block or code
```
The documents are not updated here (return error during emulation) and `UC_ERR_ARGS` seems a typo.

PhilippTakacs committed:
fixup! fixup! hooks allow to change the user data of block and code hooks
